moztelemetry/spark.py

```diff
@@ -145,6 +150,95 @@ def get_one_ping_per_client(pings):
                     reduceByKey(lambda p1, p2: p1).\
                     map(lambda p: p[1])
 
+def get_data_sources():
+    try:
+        key = _bucket_meta.get_key("sources.json")
+        sources = key.get_contents_as_string()
+        return json.loads(sources)
+    except ssl.SSLError:
+        return {}
+
+def get_source_schema(source_name):
+    global _sources
+    if _sources is None:
+        _sources = get_data_sources()
+
+    if source_name not in _sources:
+        raise ValueError("Unknown data source: {}. Known sources: [{}].".format(source_name, ", ".join(sorted(_sources.keys()))))
+
+    if "schema" not in _sources[source_name]:
+        try:
+            key = _bucket_meta.get_key("{}/schema.json".format(_sources[source_name].get("metadata_prefix", source_name)))
+            schema = key.get_contents_as_string()
+            _sources[source_name]["schema"] = json.loads(schema)
+        except ssl.SSLError:
+            return None
+    return _sources[source_name]["schema"]
+
+def get_records(sc, source_name, **kwargs):
+    schema = get_source_schema(source_name)
+    if schema is None:
+        raise ValueError("Error getting schema for {}".format(source_name))
+
+    bucket_name = _sources[source_name]["bucket"]
+    bucket_prefix = _sources[source_name]["prefix"]
+    if bucket_prefix[-1] != "/":
+        bucket_prefix = bucket_prefix + "/"
+
+    fraction = kwargs.pop("fraction", 1.0)
+    if fraction < 0 or fraction > 1:
+        raise ValueError("Invalid fraction argument")
+
+    filter_args = {}
+    field_names = [f["field_name"] for f in schema["dimensions"]]
+    for field_name in field_names:
+        filter_args[field_name] = kwargs.pop(field_name, None)
+
+    if kwargs:
+        raise TypeError("Unexpected **kwargs {}".format(repr(kwargs)))
+
+    # We should eventually support data sources in other buckets, but for now,
+    # assume that everything lives in the v4 data bucket.
+    assert(bucket_name == _bucket_v4.name)
+    # TODO: cache the buckets, or at least recognize the ones we already have
+    #       handles to (_bucket_* vars)
+    bucket = _bucket_v4 # TODO: bucket = _conn.get_bucket(bucket_name, validate=False)
+    filter_schema = _filter_to_schema(schema, filter_args)
+    files = _list_s3_filenames(bucket, bucket_prefix, filter_schema)
+
+    if files and fraction != 1.0:
+        sample = random.choice(files, size=len(files)*fraction, replace=False)
+    else:
+        sample = files
+
+    # TODO: Make sure that "bucket_name" matches the v4 bucket name, otherwise
+    #       introduce a "bucket" parameter to _read_v4_ranges
+    parallelism = max(len(sample), sc.defaultParallelism)
+    ranges = sc.parallelize(sample, parallelism).flatMap(_read_v4_ranges).collect()
+
+    if len(ranges) == 0:
+        return sc.parallelize([])
+    else:
+        return sc.parallelize(ranges, len(ranges)).flatMap(_read_v4_range)
+
+def _list_s3_filenames(bucket, prefix, schema):
+    filenames = []
```

Comment:
Can I pass an iterator to "random.choice" instead of an array? If so, I don't even need this function.